File: src/w3g_parser/decompressor.py
"""Block decompression for W3G replay files."""

import struct
import zlib
from typing import BinaryIO

from w3g_parser.exceptions import DecompressionError, TruncatedDataError
from w3g_parser.models import ReplayHeader


def decompress_blocks(stream: BinaryIO, header: ReplayHeader) -> bytes:
    """Decompress all data blocks from stream.

    Block format varies by version:

    Classic (pre-Reforged) - 8 byte header:
    - Offset 0x00: Compressed size (1 word)
    - Offset 0x02: Decompressed size (1 word)
    - Offset 0x04: Checksum/unknown (1 dword)
    - Offset 0x08: Compressed data (compressed_size bytes, raw deflate)

    Reforged - 12 byte header:
    - Offset 0x00: Compressed size (1 word) - doesn't include header
    - Offset 0x02: Unknown (1 word)
    - Offset 0x04: Decompressed size (1 dword)
    - Offset 0x08: Checksum (1 dword)
    - Offset 0x0C: Compressed data (zlib format with header)

    Args:
        stream: Binary stream positioned after header
        header: Parsed replay header

    Returns:
        Concatenated decompressed data

    Raises:
        DecompressionError: If decompression fails
        TruncatedDataError: If not enough data
    """
    decompressed_parts: list[bytes] = []

    # Determine if this is Reforged format
    # Reforged uses 12-byte block headers and zlib with headers
    is_reforged = header.is_reforged

    for block_num in range(header.num_compressed_blocks):
        if is_reforged:
            # Reforged: 12-byte header
            block_header = stream.read(12)
            if len(block_header) < 12:
                raise TruncatedDataError(
                    f"Block {block_num} header truncated",
                    stream.tell()
                )

            # Parse Reforged header
            # compressed_size is the size of zlib data only
            compressed_size = struct.unpack_from("<H", block_header, 0)[0]
            decompressed_size = struct.unpack_from("<I", block_header, 4)[0]

            # Read compressed data
            compressed_data = stream.read(compressed_size)
            if len(compressed_data) < compressed_size:
                raise TruncatedDataError(
                    f"Block {block_num} data truncated: expected {compressed_size}, "
                    f"got {len(compressed_data)}",
                    stream.tell()
                )

            # Decompress using zlib (with header)
            try:
                decompressor = zlib.decompressobj()
                decompressed = decompressor.decompress(compressed_data)
                decompressed += decompressor.flush()
            except zlib.error as e:
                raise DecompressionError(
                    f"Block {block_num} decompression failed: {e}",
                    stream.tell() - compressed_size
                ) from e

        else:
            # Classic: 8-byte header
            block_header = stream.read(8)
            if len(block_header) < 8:
                raise TruncatedDataError(
                    f"Block {block_num} header truncated",
                    stream.tell()
                )

            compressed_size, decompressed_size, _ = struct.unpack("<HHI", block_header)

            # Read compressed data
            compressed_data = stream.read(compressed_size)
            if len(compressed_data) < compressed_size:
                raise TruncatedDataError(
                    f"Block {block_num} data truncated: expected {compressed_size}, "
                    f"got {len(compressed_data)}",
                    stream.tell()
                )

            # Decompress using raw deflate (no header)
            try:
                decompressor = zlib.decompressobj(wbits=-15)
                decompressed = decompressor.decompress(compressed_data)
                decompressed += decompressor.flush()
            except zlib.error:
                # Fall back to zlib with header
                try:
                    decompressor = zlib.decompressobj()
                    decompressed = decompressor.decompress(compressed_data)
                    decompressed += decompressor.flush()
                except zlib.error as e:
                    raise DecompressionError(
                        f"Block {block_num} decompression failed: {e}",
                        stream.tell() - compressed_size
                    ) from e

        decompressed_parts.append(decompressed)

    return b"".join(decompressed_parts)


def decompress_single_block(data: bytes, use_zlib_header: bool = False) -> bytes:
    """Decompress a single block of data.

    Args:
        data: Raw block data (without header)
        use_zlib_header: If True, use zlib format; otherwise raw deflate

    Returns:
        Decompressed data
    """
    if use_zlib_header:
        decompressor = zlib.decompressobj()
    else:
        decompressor = zlib.decompressobj(wbits=-15)

    try:
        result = decompressor.decompress(data)
        result += decompressor.flush()
        return result
    except zlib.error:
        # Try the other format
        if use_zlib_header:
            decompressor = zlib.decompressobj(wbits=-15)
        else:
            decompressor = zlib.decompressobj()
        result = decompressor.decompress(data)
        result += decompressor.flush()
        return result

File: pkg/w3g/decompressor.go
package w3g

import (
	"bytes"
	"compress/flate"
	"compress/zlib"
	"encoding/binary"
	"fmt"
	"io"
)

// decompressBlocks decompresses all data blocks from reader.
//
// Block format varies by version:
//
// Classic (pre-Reforged) - 8 byte header:
//   - Offset 0x00: Compressed size (1 word)
//   - Offset 0x02: Decompressed size (1 word)
//   - Offset 0x04: Checksum/unknown (1 dword)
//   - Offset 0x08: Compressed data (compressed_size bytes, raw deflate)
//
// Reforged - 12 byte header:
//   - Offset 0x00: Compressed size (1 word) - doesn't include header
//   - Offset 0x02: Unknown (1 word)
//   - Offset 0x04: Decompressed size (1 dword)
//   - Offset 0x08: Checksum (1 dword)
//   - Offset 0x0C: Compressed data (zlib format with header)
func decompressBlocks(r io.Reader, header *ReplayHeader) ([]byte, error) {
	var result bytes.Buffer
	isReforged := header.IsReforged()

	for i := uint32(0); i < header.NumCompressedBlocks; i++ {
		var compressedSize uint32

		if isReforged {
			// Reforged: 12-byte header
			blockHeader := make([]byte, 12)
			if _, err := io.ReadFull(r, blockHeader); err != nil {
				return nil, newTruncatedDataError(
					fmt.Sprintf("block %d header truncated", i), result.Len(),
				)
			}
			compressedSize = uint32(binary.LittleEndian.Uint16(blockHeader[0:]))
			// decompressedSize is at offset 4 (4 bytes) but we don't need it
		} else {
			// Classic: 8-byte header
			blockHeader := make([]byte, 8)
			if _, err := io.ReadFull(r, blockHeader); err != nil {
				return nil, newTruncatedDataError(
					fmt.Sprintf("block %d header truncated", i), result.Len(),
				)
			}
			compressedSize = uint32(binary.LittleEndian.Uint16(blockHeader[0:]))
			// decompressedSize is at offset 2 (2 bytes) but we don't need it
		}

		// Read compressed data
		compressedData := make([]byte, compressedSize)
		if _, err := io.ReadFull(r, compressedData); err != nil {
			return nil, newTruncatedDataError(
				fmt.Sprintf("block %d data truncated: expected %d bytes", i, compressedSize),
				result.Len(),
			)
		}

		// Decompress
		var decompressed []byte
		var err error

		if isReforged {
			// Reforged uses zlib with header
			decompressed, err = decompressZlib(compressedData)
		} else {
			// Classic uses raw deflate
			decompressed, err = decompressDeflate(compressedData)
			if err != nil {
				// Fallback to zlib with header
				decompressed, err = decompressZlib(compressedData)
			}
		}

		if err != nil {
			return nil, newDecompressionError(
				fmt.Sprintf("block %d decompression failed: %v", i, err),
				result.Len(),
			)
		}

		result.Write(decompressed)
	}

	return result.Bytes(), nil
}

// decompressZlib decompresses data using zlib (with header).
func decompressZlib(data []byte) ([]byte, error) {
	r, err := zlib.NewReader(bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	defer r.Close()

	// Use a buffer to read the decompressed data
	// The W3G format doesn't always have proper zlib trailers,
	// so io.ReadAll may fail with "unexpected EOF"
	var result bytes.Buffer
	buf := make([]byte, 8192)
	for {
		n, err := r.Read(buf)
		if n > 0 {
			result.Write(buf[:n])
		}
		if err != nil {
			if err == io.EOF || err.Error() == "unexpected EOF" {
				break
			}
			// Only return error if we haven't read any data
			if result.Len() == 0 {
				return nil, err
			}
			break
		}
	}

	return result.Bytes(), nil
}

// decompressDeflate decompresses data using raw deflate (no header).
func decompressDeflate(data []byte) ([]byte, error) {
	r := flate.NewReader(bytes.NewReader(data))
	defer r.Close()

	return io.ReadAll(r)
}

// decompressSingleBlock decompresses a single block of data.
func decompressSingleBlock(data []byte, useZlibHeader bool) ([]byte, error) {
	if useZlibHeader {
		return decompressZlib(data)
	}

	decompressed, err := decompressDeflate(data)
	if err != nil {
		// Try the other format as fallback
		return decompressZlib(data)
	}
	return decompressed, nil
}
